en/news/_posts/2019-12-06-separation-of-positional-and-keyword-arguments-in-ruby-3-0.md

```diff
@@ -0,0 +1,341 @@
+
+## Handling argument delegation
+
+### Ruby 2.6 or prior
+
+In Ruby 2, you can write a delegation method by accepting a `*rest` argument and a `&block` argument, and passing the two to the target method.  In this behavior, the keyword arguments are also implicitly handled by the automatic conversion between positional and keyword arguments.
+
+```
+def foo(*args, &block)
+  target(*args, &block)
+end
+```
+
+### Ruby 3
+
+You need to explicitly delegate keyword arguments.
+
+```
+def foo(*args, **kwargs, &block)
+  target(*args, **kwargs, &block)
+end
+```
+
+Alternatively, if you do not need compatibility with Ruby 2.6 or prior and you don't alter any arguments, you can use a delegation syntax (`...`) that is introduced in Ruby 2.7.
+
+```
+def foo(...)
+  target(...)
+end
+```
+
+### Ruby 2.7
+
+In short: use `Module#ruby2_keywords` and delegate `*args, &block`.
+
+```
+ruby2_keywords def foo(*args, &block)
+  target(*args, &block)
+end
+```
+
+`ruby2_keywords` accepts keyword arguments as the last Hash argument, and passes it as keyword arguments when calling the other method.
+
+In fact, Ruby 2.7 allows the new style of delegation in many cases.  However, there is a known corner case.  See the next section.
+
+### A compatible delegation that works on Ruby 2.6, 2.7 and Ruby 3
+
+In short: use `Module#ruby2_keywords` again.
+
+```
+ruby2_keywords def foo(*args, &block)
+  target(*args, &block)
+end
+```
+
+Unfortunately, we need to use the old-style delegation (i.e., no `**kwargs`) because Ruby 2.6 or prior does not handle the new delegation style correctly.  This is one of the reasons of the keyword argument separation; the details are described in the final section. And `ruby2_keywords` allows you to run the old style even in Ruby 2.7 and 3.0.  As there is no `ruby2_keywords` defined in 2.6 or prior, please use [ruby2_keywords gem](https://rubygems.org/gems/ruby2_keywords) or define it by yourself:
+
+```
+def ruby2_keywords(*)
+end if RUBY_VESRION < "2.7"
+```
+
+---
+
+If your code doesn't have to run on Ruby 2.6 or older, and you don't care a corner case, you may try the new style in Ruby 2.7.  In almost all cases, it works.  Note that, however, there are unfortunate corner cases as follows:
+
+```
+def target(*args)
+  p args
+end
+
+def foo(*args, **kwargs, &block)
+  target(*args, **kwargs, &block)
+end
+
+foo({})       #=> Ruby 2.7: []   ({} is dropped)
+foo({}, **{}) #=> Ruby 2.7: [{}] (You can pass {} by explicitly passing "no" keywords)
+```
+
+An empty Hash argument is automatically converted and absorbed to `**kwargs`, and the delegation call removes the empty keyword hash, so no argument is passed to `target`.  As far as we know, this is the only corner case.
+
+As noted in the last line, you can work around this issue by using `**{}`.
+
+If you really worry about the portability, use `ruby2_keywords`.  (Acknowledge that Ruby 2.6 or before themselves have tons of corner cases in keyword arguments. :-)
+
```

Comment:
```suggestion
`ruby2_keywords` might be removed in the future after Ruby 2.6 reaches end-of-life. At that point, we recommend to explicitly delegate keyword arguments (see Ruby 3 code above). If you still need to support Ruby 2.6 at that time, you will need a version check and use a mix of both approaches.

```

Reply:
I agree with the first two sentences.  The last one is to allow users to use 2.6 even after EOL?  It is non-sense to me.

Reply:
> `ruby2_keywords` might be removed in the future after Ruby 2.6 reaches end-of-life.
> At that point, we recommend to explicitly delegate keyword arguments (see Ruby 3 code above).

Third sentence:
> If you still need to support Ruby 2.6 at that time, you will need a version check and use a mix of both approaches.

I mean if some gem wants to support some Ruby 3.x which would not have `ruby2_keywords` *and* Ruby 2.6, they will need a version check like:
```ruby
if RUBY_VERSION >= "3.0" # or if !defined?(ruby2_keywords) if we are careful with the ruby2_keywords gem.
  def foo(*args, **kwargs, &block)
    target(*args, **kwargs, &block)
  end
else
  ruby2_keywords def foo(*args, &block)
    target(*args, &block)
  end
end
```

Many gems support Ruby versions far after the corresponding CRuby version is officially EOL, so we cannot ignore that case.

Reply:
I'm unsure if I meant what I'd like to say... Of course, people can use 2.6 after EOL, at their own risk.  I think it is non-sense for us (Ruby core team) to care and mention such a case.

Reply:
I think we need to care a bit about that case, because the kwargs changes are going to be pretty ugly for those cases. I think it's better to write something than nothing which might be interpreted as "the Ruby core team doesn't care about that case and their only written solution is to drop Ruby 2.6 support, which might not be an option for that gem".

Could you approve the suggestion and then adapt the third sentence in a next commit?

Reply:
I don't think we want to mention version checks as a way to keep far future and 2.6 compatibility.

* Every place you are mentioning a version check will result in slower code in CRuby 3.0 (unless someone rewrites CRuby's keyword support). 
* We don't know what the status will be with `ruby2_keywords`, but it is likely it will be present until at least 2.6 EOL, which means Ruby 3.2 will still support it.  We can reevaluate later whether it should be kept longer.
* We shouldn't go out of our way to cater to people who use EOL versions.
* If we do drop `ruby2_keywords`, we can have a new (and much smaller) document that shows users how to cope with that change.

Reply:
> Could you approve the suggestion and then adapt the third sentence in a next commit?

Done, and removed the third sentence.

Reply:
> Every place you are mentioning a version check will result in slower code in CRuby 3.0 (unless someone rewrites CRuby's keyword support).

That depends on the check. If it was something like `if RUBY_VERSION > "3.2"` or `if !defined?(ruby2_keywords)` then it wouldn't slow down code for any of those version before `ruby2_keywords` is removed (that needs some care in the `ruby2_keywords` gem to only define the `ruby2_keywords` method if < 2.7, it would anyway be incorrect for later versions, I'll make a PR).
After `ruby2_keywords` is removed it can't be used obviously, and I'd say the earlier the better since we'll have to face it eventually anyway.

I'm fine to not have code about the version check in the blog post.
Just the third sentence above is better than nothing.

FWIW, I still consider it's a huge hack we need a magic `ruby2_keywords` modifier to get delegation working in a programming language, and that we broke `*args`-delegation. The blog post already mentions:
```
def foo(*args, **kwargs, &block)
  target(*args, **kwargs, &block)
end
```
as it should, because that is the logical/clean/intuitive way to do delegation in Ruby 3+.
(I'm also fine with new ways like `...`)

IMHO we should focus on making those fast, because those ways are clean and intuitive.
`ruby2_keywords` won't be forever, I think it makes very little sense to use `ruby2_keywords` in new code after the Ruby 3.0 release and IMHO a hack should not be used just for performance of delegation.

Back to this blog post, from what I see you're not against my original suggestion, but rather about the specifics of the code above. So can we just apply my suggestion and propose further tweaks if desired?

Reply:
> Done, and removed the third sentence.

OK, better than nothing, thanks.